{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# reference fasta libraries:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2016-06-28T10:42:59.165964",
     "start_time": "2016-06-28T10:42:58.942360"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "565\n",
      "2588\n",
      "672177\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd ~/Downloads/base\n",
    "mkdir -p test_smallonly\n",
    "cd ~/Downloads/base/test_smallonly\n",
    "rm -rf /home/guillaume/Downloads/base/test_smallonly/*\n",
    "\n",
    "# references:\n",
    "hairpin='/home/guillaume/Homo_sapiens_UCSC_hg38/hairpin_hsa.fa'\n",
    "rRNA='/home/guillaume/Downloads/base/israp/rRNA.fa'\n",
    "mature_hsa_filtered_T=\"/home/guillaume/Homo_sapiens_UCSC_hg38/mature_hsa_filtered_T.fa\"\n",
    "smallRNA=\"/home/guillaume/Homo_sapiens_UCSC_hg38/smallRNA.fa\"\n",
    "grep \">\" $rRNA | wc -l\n",
    "grep \">\" $mature_hsa_filtered_T | wc -l\n",
    "grep \">\" $smallRNA | wc -l\n",
    "\n",
    "# choose reference:\n",
    "#cat $hairpin $rRNA $mature_hsa_filtered_T $smallRNA > combined.fasta\n",
    "cat $mature_hsa_filtered_T > combined.fasta\n",
    "cp combined.fasta reference.fasta\n",
    "\n",
    "# if necessary, replace U with T in .fa file\n",
    "sed -i s/U/T/g ./reference.fasta\n",
    "\n",
    "# if necessary, replace lower case atgc with uppercase ATGC in .fa file\n",
    "awk '{ if ($0 !~ />/) {print toupper($0)} else {print $0} }' reference.fasta > reference2.fasta\n",
    "\n",
    "# if necessary, replace Y with N \n",
    "perl -e 'while(<>) { if ($_ =~ /^>.*/) { print $_; } else { $_ =~ tr/Y/N/; print $_;}}' < reference2.fasta > reference.fasta\n",
    "\n",
    "/home/guillaume/.local/bin/fasta_formatter -i reference2.fasta -o reference.oneline.fasta -w 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2016-06-28T10:43:04.913235",
     "start_time": "2016-06-28T10:43:04.886105"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# collapse duplicate sequences in reference:\n",
    "#/home/guillaume/.local/bin/fastx_collapser < reference.oneline.fasta > reference.collapsed.fasta\n",
    "inname='/home/guillaume/Downloads/base/test_smallonly/reference.oneline.fasta'\n",
    "oname='/home/guillaume/Downloads/base/test_smallonly/reference.collapsed.fasta'\n",
    "\n",
    "from itertools import groupby\n",
    "\n",
    "if __name__ == '__main__':\n",
    "\n",
    "    ishead = lambda x: x.startswith('>')\n",
    "    all_seqs = set()\n",
    "    with open(inname) as handle:\n",
    "        with open(oname, 'w') as outhandle:\n",
    "            head = None\n",
    "            for h, lines in groupby(handle, ishead):\n",
    "                if h:\n",
    "                    head = lines.__next__()\n",
    "                else:\n",
    "                    seq = ''.join(lines)\n",
    "                    if seq not in all_seqs:\n",
    "                        all_seqs.add(seq)\n",
    "                        outhandle.write('%s%s' % (head, seq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2016-06-28T10:43:07.606556",
     "start_time": "2016-06-28T10:43:06.957717"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2565\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd ~/Downloads/base/test_smallonly\n",
    "cp reference.collapsed.fasta reference.fasta \n",
    "/home/guillaume/.local/bin/fastx_quality_stats -i reference.fasta -o reference.fasta.quality.txt\n",
    "/home/guillaume/.local/bin/fastx_nucleotide_distribution_graph.sh -i reference.fasta.quality.txt -o reference.distrib.png\n",
    "\n",
    "# report info\n",
    "grep \">\" reference.fasta | wc -l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2016-06-28T10:43:09.564329",
     "start_time": "2016-06-28T10:43:08.999450"
    },
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "column\tcount\tmin\tmax\tsum\tmean\tQ1\tmed\tQ3\tIQR\tlW\trW\tA_Count\tC_Count\tG_Count\tT_Count\tN_Count\tMax_count\n",
      "1\t2565\t100\t-100\t0\t0.00\t52\t260\t367\t315\t100\t-100\t719\t556\t358\t932\t0\t2565\n",
      "2\t2565\t100\t-100\t0\t0.00\t52\t260\t262\t210\t100\t-100\t684\t570\t743\t568\t0\t2565\n",
      "3\t2565\t100\t-100\t0\t0.00\t52\t260\t262\t210\t100\t-100\t665\t542\t790\t568\t0\t2565\n",
      "4\t2565\t100\t-100\t0\t0.00\t155\t262\t262\t107\t100\t-100\t544\t603\t939\t479\t0\t2565\n",
      "5\t2565\t100\t-100\t0\t0.00\t155\t262\t262\t107\t100\t-100\t562\t547\t898\t558\t0\t2565\n",
      "6\t2565\t100\t-100\t0\t0.00\t155\t260\t262\t107\t100\t-100\t574\t657\t738\t596\t0\t2565\n",
      "7\t2565\t100\t-100\t0\t0.00\t52\t260\t262\t210\t100\t-100\t643\t619\t713\t590\t0\t2565\n",
      "8\t2565\t100\t-100\t0\t0.00\t155\t260\t365\t210\t100\t-100\t575\t649\t689\t652\t0\t2565\n",
      "9\t2565\t100\t-100\t0\t0.00\t157\t262\t365\t208\t100\t-100\t533\t621\t691\t720\t0\t2565\n",
      "10\t2565\t100\t-100\t0\t0.00\t157\t262\t365\t208\t100\t-100\t519\t638\t756\t652\t0\t2565\n",
      "11\t2565\t100\t-100\t0\t0.00\t155\t262\t262\t107\t100\t-100\t549\t602\t820\t594\t0\t2565\n",
      "12\t2565\t100\t-100\t0\t0.00\t155\t262\t365\t210\t100\t-100\t589\t580\t717\t679\t0\t2565\n",
      "13\t2565\t100\t-100\t0\t0.00\t155\t262\t365\t210\t100\t-100\t544\t548\t769\t704\t0\t2565\n",
      "14\t2565\t100\t-100\t0\t0.00\t155\t262\t365\t210\t100\t-100\t542\t566\t751\t706\t0\t2565\n",
      "15\t2565\t100\t-100\t0\t0.00\t155\t262\t365\t210\t100\t-100\t557\t587\t775\t646\t0\t2565\n",
      "16\t2565\t100\t-100\t0\t0.00\t157\t260\t365\t208\t100\t-100\t510\t695\t676\t684\t0\t2565\n",
      "17\t2557\t100\t-100\t0\t0.00\t157\t260\t365\t208\t100\t-100\t496\t697\t710\t654\t0\t2565\n",
      "18\t2497\t100\t-100\t0\t0.00\t157\t262\t365\t208\t100\t-100\t511\t599\t694\t693\t0\t2565\n",
      "19\t2421\t100\t-100\t0\t0.00\t157\t260\t365\t208\t100\t-100\t488\t632\t642\t659\t0\t2565\n",
      "20\t2324\t100\t-100\t0\t0.00\t155\t262\t365\t210\t100\t-100\t495\t516\t710\t603\t0\t2565\n",
      "21\t2162\t100\t-100\t0\t0.00\t155\t262\t262\t107\t100\t-100\t460\t482\t682\t538\t0\t2565\n",
      "22\t1649\t100\t-100\t0\t0.00\t155\t262\t365\t210\t100\t-100\t347\t371\t465\t466\t0\t2565\n",
      "23\t515\t100\t-100\t0\t0.00\t155\t260\t365\t210\t100\t-100\t99\t93\t165\t158\t0\t2565\n",
      "24\t140\t100\t-100\t0\t0.00\t50\t50\t155\t105\t100\t-100\t25\t28\t50\t37\t0\t2565\n",
      "25\t44\t100\t-100\t0\t0.00\t50\t50\t50\t0\t100\t-100\t8\t7\t12\t17\t0\t2565\n",
      "26\t12\t100\t-100\t0\t0.00\t50\t50\t50\t0\t100\t-100\t4\t3\t3\t2\t0\t2565\n",
      "27\t4\t100\t-100\t0\t0.00\t50\t50\t50\t0\t100\t-100\t2\t1\t1\t0\t0\t2565\n",
      "28\t1\t100\t-100\t0\t0.00\t100\t100\t100\t0\t100\t-100\t0\t0\t1\t0\t0\t2565\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[bwa_index] Pack FASTA... 0.01 sec\n",
      "[bwa_index] Construct BWT for the packed sequence...\n",
      "[bwa_index] 0.02 seconds elapse.\n",
      "[bwa_index] Update BWT... 0.00 sec\n",
      "[bwa_index] Pack forward-only FASTA... 0.01 sec\n",
      "[bwa_index] Construct SA from BWT and Occ... 0.01 sec\n",
      "[main] Version: 0.7.12-r1039\n",
      "[main] CMD: /home/guillaume/Downloads/bwa.kit/bwa index reference.fasta\n",
      "[main] Real time: 0.465 sec; CPU: 0.040 sec\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd ~/Downloads/base/test_smallonly\n",
    "/home/guillaume/.local/bin/fastx_quality_stats -i reference.fasta\n",
    "# if necessary, index reference:\n",
    "~/Downloads/bwa.kit/bwa index reference.fasta\n",
    "##bowtie2-build reference.fasta reference"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "choose sample, remove adapters and select read length range:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# merge PEAR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2016-06-28T10:57:59.733456",
     "start_time": "2016-06-28T10:45:02.548967"
    },
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Process is interrupted.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd ~/Downloads/base\n",
    "mkdir -p test_smallonly\n",
    "mkdir -p test_smallonly/peared\n",
    "rm -rf ./test_smallonly/peared/*\n",
    "\n",
    "# select sample:\n",
    "samples=\"RF-118 RF-115 RF-117 RF-59 RF-66 RF67 RF55-1 RF112 RF111 RF105 RF104 RF79 RF1 106-1 109-2 110-3 113 107 108-1 RF-102 RF-85D RF-81 RF-80 RF-62 RF-50 RF-49 RF-42 RF-32 RF-31 RF-30 RF-20 RF-17 RF-16 RF-12 RF-10 RF-5 RF-4\"\n",
    "\n",
    "# remove adapter and trim reads:\n",
    "for sample in $samples; do\n",
    "    echo $sample\n",
    "    fileR1=$(ls | grep .*${sample}_.*R1.*)\n",
    "    fileR2=$(ls | grep .*${sample}_.*R2.*)\n",
    "    echo $fileR1\n",
    "    echo $fileR2\n",
    "\n",
    "~/.local/bin/cutadapt -f fastq -e 0.1 -q 20,20 -O 1 -m 16 -M 200 --max-n 2 --discard-untrimmed --match-read-wildcards \\\n",
    "-a AGATCGGAAGAGCACACGTCT -A GATCGTCGGACTGTAGAACTCTGAAC \\\n",
    "-o ./cutadapt_paired/$fileR1 -p ./cutadapt_paired/$fileR2 $fileR1 $fileR2\n",
    "\n",
    "pear -f ./cutadapt_paired/$fileR1 -r ./cutadapt_paired/$fileR2 -o ./test_smallonly/peared/$fileR1 -j 4 -u 1 -y 10G -n 0\n",
    "    rm ./test_smallonly/peared/*unassembled*\n",
    "    rm ./test_smallonly/peared/*discarded*\n",
    "    done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# align"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2016-06-28T13:21:41.337767",
     "start_time": "2016-06-28T11:03:05.142693"
    },
    "code_folding": [],
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "[bam_sort_core] merging from 3 files...\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "[bam_sort_core] merging from 2 files...\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "[bam_sort_core] merging from 3 files...\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "[bam_sort_core] merging from 2 files...\n",
      "[bam_sort_core] merging from 2 files...\n",
      "[bam_sort_core] merging from 2 files...\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "[bam_sort_core] merging from 2 files...\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "[bam_sort_core] merging from 5 files...\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "[bam_sort_core] merging from 2 files...\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "[bam_sort_core] merging from 2 files...\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "[bam_sort_core] merging from 2 files...\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "[bam_sort_core] merging from 2 files...\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "[bam_sort_core] merging from 2 files...\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "[bam_sort_core] merging from 2 files...\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "[bam_sort_core] merging from 2 files...\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "[bam_sort_core] merging from 3 files...\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "[bam_sort_core] merging from 2 files...\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "[bam_sort_core] merging from 2 files...\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "[bam_sort_core] merging from 2 files...\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "[bam_sort_core] merging from 2 files...\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "[bam_sort_core] merging from 2 files...\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "[bam_sort_core] merging from 2 files...\n",
      "[bam_sort_core] merging from 4 files...\n",
      "[bam_sort_core] merging from 3 files...\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "[bam_sort_core] merging from 3 files...\n",
      "[bam_sort_core] merging from 3 files...\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "[bam_sort_core] merging from 3 files...\n",
      "[bam_sort_core] merging from 3 files...\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "[bam_sort_core] merging from 3 files...\n",
      "[bam_sort_core] merging from 11 files...\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "[bam_sort_core] merging from 2 files...\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "[bam_sort_core] merging from 2 files...\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "[bam_sort_core] merging from 2 files...\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "[bam_sort_core] merging from 2 files...\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence\n",
      "Reverse and complement a nucleotide sequence**OUTPUT MUTED**"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd ~/Downloads/base/test_smallonly\n",
    "\n",
    "# select sample:\n",
    "samples=\"RF-118 RF-115 RF-117 RF-59 RF-66 RF67 RF55-1 RF112 RF111 RF105 RF104 RF79 RF1 106-1 109-2 110-3 113 107 108-1 RF-102 RF-85D RF-81 RF-80 RF-62 RF-50 RF-49 RF-42 RF-32 RF-31 RF-30 RF-20 RF-17 RF-16 RF-12 RF-10 RF-5 RF-4\"\n",
    "#samples=\"RF-118 RF-115\"\n",
    "\n",
    "mkdir -p bams\n",
    "rm -rf ./bams/*\n",
    "mkdir -p counts\n",
    "rm -rf ./counts/*\n",
    "mkdir -p stat_mapped\n",
    "rm -rf ./stat_mapped/*\n",
    "\n",
    "for sample in $samples; do\n",
    "    fileR1=$(ls ./peared/ | grep ${sample}_.*R1.*)\n",
    "    fileR2=$(ls ./peared/ | grep ${sample}_.*R2.*)\n",
    "    \n",
    "    ~/Downloads/bwa.kit/bwa aln -n 1 -o 0 -e 0 -k 1 -t 4 reference.fasta ./peared/$fileR1 > R1.sai 2> error\n",
    "    ~/Downloads/bwa.kit/bwa samse reference.fasta R1.sai ./peared/$fileR1 > $fileR1.sam 2> error\n",
    "    # sam to bam \n",
    "    samtools view -Sb $fileR1.sam > view.bam\n",
    "    mv view.bam $fileR1.bam\n",
    "    # bam to sorted bam\n",
    "    samtools sort $fileR1.bam -o $fileR1.sorted.bam\n",
    "    # sorted bam to index sorted bam\n",
    "    samtools index $fileR1.sorted.bam\n",
    "    # convert to bed\n",
    "    bamToBed -i $fileR1.sorted.bam > $fileR1.sorted.bed\n",
    "    # generate statistics: percentage of patients reads mapped to genome hg38\n",
    "    samtools flagstat $fileR1.sorted.bam > $fileR1.percentage_mapped.txt\n",
    "    samtools idxstats $fileR1.sorted.bam > $fileR1.count.txt\n",
    "      \n",
    "    ### now trim 1 base of unmapped reads\n",
    "    samtools view -b -f 4 $fileR1.sorted.bam > unmapped.bam\n",
    "    samtools sort -n -O BAM -o aln.qsort.bam unmapped.bam \n",
    "    bedtools bamtofastq -i aln.qsort.bam -fq unmapped.fq\n",
    "    cat unmapped.fq | paste - - - - | cut -f1-2 | sed 's/^@/>/g' | tr '\\t' '\\n' > unmapped.fasta\n",
    "    revseq unmapped.fasta -reverse -nocomplement -outseq  unmapped.rev.fasta\n",
    "    /home/guillaume/.local/bin/fasta_formatter -i unmapped.rev.fasta -o unmapped.rev.oneline.fasta -w 0\n",
    "    /home/guillaume/.local/bin/fastx_trimmer -f 2 -i unmapped.rev.oneline.fasta -o unmapped.rev.trimmed1.fasta\n",
    "    revseq unmapped.rev.trimmed1.fasta -reverse -nocomplement -outseq  unmapped.trimmed1.fasta\n",
    "    /home/guillaume/.local/bin/fasta_formatter -i unmapped.trimmed1.fasta -o unmapped.trimmed1.ol.fasta -w 0  \n",
    "    sed -n '1~2p' unmapped.fasta > firstline\n",
    "    sed -n '0~2p' unmapped.trimmed1.ol.fasta > secondline\n",
    "    paste -d\"\\n\" firstline secondline > $fileR1.unmapped.trimmed1.fasta\n",
    "    \n",
    "    ## align trimmed1\n",
    "    ~/Downloads/bwa.kit/bwa aln -n 1 -o 0 -e 0 -k 1 -t 4 reference.fasta $fileR1.unmapped.trimmed1.fasta > R1.sai 2> error\n",
    "    ~/Downloads/bwa.kit/bwa samse reference.fasta R1.sai $fileR1.unmapped.trimmed1.fasta > $fileR1.t1.sam 2> error\n",
    "    # sam to bam \n",
    "    samtools view -Sb $fileR1.t1.sam > view.bam\n",
    "    mv view.bam $fileR1.t1.bam\n",
    "    # bam to sorted bam\n",
    "    samtools sort $fileR1.t1.bam -o $fileR1.sorted.t1.bam\n",
    "    # sorted bam to index sorted bam\n",
    "    samtools index $fileR1.sorted.t1.bam\n",
    "    # convert to bed\n",
    "    bamToBed -i $fileR1.sorted.t1.bam > $fileR1.sorted.t1.bed\n",
    "    # generate statistics: percentage of patients reads mapped to genome hg38\n",
    "    samtools flagstat $fileR1.sorted.t1.bam > $fileR1.percentage_mapped.t1.txt\n",
    "    samtools idxstats $fileR1.sorted.t1.bam > $fileR1.count.t1.txt   \n",
    "    \n",
    "        ### now trim 1 base of unmapped reads\n",
    "    samtools view -b -f 4 $fileR1.sorted.t1.bam > unmapped.bam\n",
    "    samtools sort -n -O BAM -o aln.qsort.bam unmapped.bam \n",
    "    bedtools bamtofastq -i aln.qsort.bam -fq unmapped.fq\n",
    "    cat unmapped.fq | paste - - - - | cut -f1-2 | sed 's/^@/>/g' | tr '\\t' '\\n' > unmapped.fasta\n",
    "    revseq unmapped.fasta -reverse -nocomplement -outseq  unmapped.rev.fasta\n",
    "    /home/guillaume/.local/bin/fasta_formatter -i unmapped.rev.fasta -o unmapped.rev.oneline.fasta -w 0\n",
    "    /home/guillaume/.local/bin/fastx_trimmer -f 2 -i unmapped.rev.oneline.fasta -o unmapped.rev.trimmed1.fasta\n",
    "    revseq unmapped.rev.trimmed1.fasta -reverse -nocomplement -outseq  unmapped.trimmed1.fasta\n",
    "    /home/guillaume/.local/bin/fasta_formatter -i unmapped.trimmed1.fasta -o unmapped.trimmed1.ol.fasta -w 0  \n",
    "    sed -n '1~2p' unmapped.fasta > firstline\n",
    "    sed -n '0~2p' unmapped.trimmed1.ol.fasta > secondline\n",
    "    paste -d\"\\n\" firstline secondline > $fileR1.unmapped.trimmed2.fasta\n",
    "    \n",
    "    ## align trimmed2\n",
    "    ~/Downloads/bwa.kit/bwa aln -n 1 -o 0 -e 0 -k 1 -t 4 reference.fasta $fileR1.unmapped.trimmed2.fasta > R1.sai 2> error\n",
    "    ~/Downloads/bwa.kit/bwa samse reference.fasta R1.sai $fileR1.unmapped.trimmed2.fasta > $fileR1.t2.sam 2> error\n",
    "    # sam to bam \n",
    "    samtools view -Sb $fileR1.t2.sam > view.bam\n",
    "    mv view.bam $fileR1.t2.bam\n",
    "    # bam to sorted bam\n",
    "    samtools sort $fileR1.t2.bam -o $fileR1.sorted.t2.bam\n",
    "    # sorted bam to index sorted bam\n",
    "    samtools index $fileR1.sorted.t2.bam\n",
    "    # convert to bed\n",
    "    bamToBed -i $fileR1.sorted.t2.bam > $fileR1.sorted.t2.bed\n",
    "    # generate statistics: percentage of patients reads mapped to genome hg38\n",
    "    samtools flagstat $fileR1.sorted.t2.bam > $fileR1.percentage_mapped.t2.txt\n",
    "    samtools idxstats $fileR1.sorted.t2.bam > $fileR1.count.t2.txt   \n",
    "\n",
    "\n",
    "    ### now trim 1 base of unmapped reads\n",
    "    samtools view -b -f 4 $fileR1.sorted.t2.bam > unmapped.bam\n",
    "    samtools sort -n -O BAM -o aln.qsort.bam unmapped.bam \n",
    "    bedtools bamtofastq -i aln.qsort.bam -fq unmapped.fq\n",
    "    cat unmapped.fq | paste - - - - | cut -f1-2 | sed 's/^@/>/g' | tr '\\t' '\\n' > unmapped.fasta\n",
    "    revseq unmapped.fasta -reverse -nocomplement -outseq  unmapped.rev.fasta\n",
    "    /home/guillaume/.local/bin/fasta_formatter -i unmapped.rev.fasta -o unmapped.rev.oneline.fasta -w 0\n",
    "    /home/guillaume/.local/bin/fastx_trimmer -f 2 -i unmapped.rev.oneline.fasta -o unmapped.rev.trimmed1.fasta\n",
    "    revseq unmapped.rev.trimmed1.fasta -reverse -nocomplement -outseq  unmapped.trimmed1.fasta\n",
    "    /home/guillaume/.local/bin/fasta_formatter -i unmapped.trimmed1.fasta -o unmapped.trimmed1.ol.fasta -w 0  \n",
    "    sed -n '1~2p' unmapped.fasta > firstline\n",
    "    sed -n '0~2p' unmapped.trimmed1.ol.fasta > secondline\n",
    "    paste -d\"\\n\" firstline secondline > $fileR1.unmapped.trimmed3.fasta\n",
    "    \n",
    "    ## align trimmed3\n",
    "    ~/Downloads/bwa.kit/bwa aln -n 1 -o 0 -e 0 -k 1 -t 4 reference.fasta $fileR1.unmapped.trimmed3.fasta > R1.sai 2> error\n",
    "    ~/Downloads/bwa.kit/bwa samse reference.fasta R1.sai $fileR1.unmapped.trimmed3.fasta > $fileR1.t3.sam 2> error\n",
    "    # sam to bam \n",
    "    samtools view -Sb $fileR1.t3.sam > view.bam\n",
    "    mv view.bam $fileR1.t3.bam\n",
    "    # bam to sorted bam\n",
    "    samtools sort $fileR1.t3.bam -o $fileR1.sorted.t3.bam\n",
    "    # sorted bam to index sorted bam\n",
    "    samtools index $fileR1.sorted.t3.bam\n",
    "    # convert to bed\n",
    "    bamToBed -i $fileR1.sorted.t3.bam > $fileR1.sorted.t3.bed\n",
    "    # generate statistics: percentage of patients reads mapped to genome hg38\n",
    "    samtools flagstat $fileR1.sorted.t3.bam > $fileR1.percentage_mapped.t3.txt\n",
    "    samtools idxstats $fileR1.sorted.t3.bam > $fileR1.count.t3.txt   \n",
    "\n",
    "    ## merge the 4 bam files per sample\n",
    "    samtools merge -f $fileR1.full.bam $fileR1.sorted.bam $fileR1.sorted.t1.bam $fileR1.sorted.t2.bam $fileR1.sorted.t3.bam\n",
    "    samtools sort $fileR1.full.bam -o $fileR1.full.sorted.bam\n",
    "    samtools index $fileR1.full.sorted.bam\n",
    "    mv $fileR1.full.sorted.bam.bai ./bams/ \n",
    "    mv $fileR1.full.sorted.bam ./bams/\n",
    "    samtools idxstats  ./bams/$fileR1.full.sorted.bam > ./counts/$fileR1.full.count.txt\n",
    "    mv *percentage* ./stat_mapped/\n",
    "    \n",
    "    done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  },
  "toc": {
   "toc_cell": false,
   "toc_number_sections": true,
   "toc_threshold": 6,
   "toc_window_display": true
  },
  "toc_position": {
   "height": "717px",
   "left": "-0.25px",
   "right": "1454.08px",
   "top": "57.95px",
   "width": "306px"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
